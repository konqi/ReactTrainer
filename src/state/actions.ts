export enum ActionType {
  ADD_TRAINEE
}
