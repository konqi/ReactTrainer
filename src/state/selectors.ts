import {ApplicationState} from '.'

export const selectTraineeById = (state: ApplicationState, traineeId: string) =>
  state.trainees[traineeId]
