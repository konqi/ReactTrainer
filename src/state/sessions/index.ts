export {
  createAddSessionAction as createAddTrainingAction,
} from './sessionActions'
export {initialSessionState} from './sessionReducer'
