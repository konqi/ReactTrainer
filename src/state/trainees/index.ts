export {reduceTrainees, initialTraineeState} from './traineeReducer'
export {
  TraineeActions,
  createIngestTraineeAction as createAddTraineeAction,
} from './traineeActions'
