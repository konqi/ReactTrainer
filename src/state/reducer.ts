import { ActionType } from "./actions";

export interface Action {
  type: ActionType;
  payload: Trainee;
}

export interface Trainee {
  name: string;
  price: number;
}

export interface ApplicationState {
  trainees: Trainee[];
}

export const initialApplicationState: ApplicationState = {
  trainees: []
};

export const reducer = (
  state: ApplicationState,
  action: Action
): ApplicationState => {
  switch (action.type) {
    case ActionType.ADD_TRAINEE:
      const newTrainees = [...state.trainees, action.payload];
      return { ...state, trainees: newTrainees };
    default:
      throw new Error();
  }
};
