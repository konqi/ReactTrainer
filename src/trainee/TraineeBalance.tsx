import {
  Container,
  createStyles,
  makeStyles,
  Paper,
  Theme,
  Typography,
} from '@material-ui/core'
import {green, red} from '@material-ui/core/colors'
import get from 'lodash/get'
import React, {useMemo} from 'react'
import {useSelector} from 'react-redux'
import {ApplicationState} from '../state'
import {Session} from '../types/Session'
import {Trainee} from '../types/Trainee'

interface ExternalProps {
  traineeId: string
}
export const TraineeBalanceConnected: React.FC<ExternalProps> = ({
  traineeId,
}: ExternalProps) => {
  const trainee = useSelector<ApplicationState, Trainee>(
    state => state.trainees[traineeId]
  )
  const sessions = useSelector<ApplicationState, Session[]>(state => {
    const sessionRefs = get(trainee, 'sessionsRef')
    if (sessionRefs !== undefined && sessionRefs.length > 0) {
      return sessionRefs.map(sessionRef => state.sessions[sessionRef])
    } else {
      return []
    }
  })

  const totalBalance = useMemo(
    () =>
      sessions.reduce(
        (prev, session) =>
          session.payedAmount - (session.price || trainee.price) + prev,
        0
      ),
    [sessions, trainee]
  )

  return <TraineeBalance totalBalance={totalBalance} />
}

const useStyles = makeStyles((theme: Theme) =>
  createStyles({
    root: {
      margin: theme.spacing(1, 0),
    },
    negativeBalance: {
      color: red[400],
    },
    positiveBalance: {
      color: green[400],
    },
  })
)

export const TraineeBalance: React.FC<{totalBalance: number}> = ({
  totalBalance,
}) => {
  const classes = useStyles()

  const balanceClass = (balance: number) => {
    if (balance > 0) {
      return classes.positiveBalance
    } else if (balance < 0) {
      return classes.negativeBalance
    }
  }

  return (
    <Paper className={classes.root}>
      <Container>
        <Typography variant="h5">
          Kontostand:&nbsp;
          <span className={balanceClass(totalBalance)}>{totalBalance} â‚¬</span>
        </Typography>
      </Container>
    </Paper>
  )
}
