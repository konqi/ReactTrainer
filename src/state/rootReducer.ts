import {ApplicationState} from '.'
import {FSA} from '../types/FSA'
import {reduceTrainees} from './trainees'

export const rootReducer = (
  state: ApplicationState,
  action: FSA<any, any>
): ApplicationState => ({
  trainees: reduceTrainees(state.trainees, action),
})
