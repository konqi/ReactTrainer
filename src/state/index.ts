import {TraineeState, TraineeFSAs} from './trainees/types'
import {initialTraineeState} from './trainees'
import {UiState, UiFSAs} from './ui/types'
import {TrainingFSAs} from './trainings/types'
import {initialUiState} from './ui/uiReducer'
import {createStore, applyMiddleware} from 'redux'
import {rootReducer} from './rootReducer'
import createSagaMiddleware from 'redux-saga'
import {rootSaga} from './sagas/rootSaga'

export interface ApplicationState {
  trainees: TraineeState
  ui: UiState
}

export const initialApplicationState: ApplicationState = {
  trainees: initialTraineeState,
  ui: initialUiState,
}

export type Action = TraineeFSAs | TrainingFSAs | UiFSAs

const sagaMiddleware = createSagaMiddleware()

export const store = createStore(
  rootReducer,
  initialApplicationState,
  applyMiddleware(sagaMiddleware)
)

sagaMiddleware.run(rootSaga)
