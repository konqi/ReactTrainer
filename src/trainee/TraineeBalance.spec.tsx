import React from 'react'
import {render, cleanup} from '@testing-library/react'
import {TraineeBalance, TraineeBalanceConnected} from './TraineeBalance'
import {Provider} from 'react-redux'
import {
  TraineeBuilder,
  SessionBuilder,
  createMockStore,
} from '../__mocks__/store'
import {initialApplicationState} from '../state'

describe('snapshot tests', () => {
  afterEach(cleanup)

  test.each`
    description   | balance
    ${'positive'} | ${1}
    ${'negative'} | ${-1}
    ${'zero'}     | ${0}
  `('renders properly with $description balance', ({balance}) => {
    const {baseElement} = render(<TraineeBalance totalBalance={balance} />)
    expect(baseElement).toMatchSnapshot()
  })
})

describe('integration tests', () => {
  afterEach(cleanup)

  test.each`
    description                     | traineePrice | sessionPrice | payedAmount | expectedBalance
    ${'zero balance'}               | ${10}        | ${undefined} | ${10}       | ${0}
    ${'underpayed'}                 | ${10}        | ${undefined} | ${5}        | ${-5}
    ${'overpayed'}                  | ${10}        | ${undefined} | ${15}       | ${5}
    ${'custom price: zero balance'} | ${10}        | ${15}        | ${15}       | ${0}
    ${'custom price: underpayed'}   | ${10}        | ${15}        | ${10}       | ${-5}
    ${'custom price: overpayed'}    | ${10}        | ${5}         | ${10}       | ${5}
  `(
    'single session tests with $description',
    ({traineePrice, sessionPrice, payedAmount, expectedBalance}) => {
      const trainee = new TraineeBuilder().withPrice(traineePrice).build()
      const session = new SessionBuilder(trainee.id)
        .withPrice(sessionPrice)
        .withPayedAmount(payedAmount)
        .build()
      trainee.sessionsRef = [session.id]
      const store = createMockStore({
        ...initialApplicationState,
        trainees: {
          [trainee.id]: trainee,
        },
        sessions: {[session.id]: session},
      })

      const {queryByText} = render(
        <Provider store={store}>
          <TraineeBalanceConnected traineeId={trainee.id} />
        </Provider>
      )

      expect(queryByText(/Kontostand/)!.firstElementChild!.textContent).toEqual(
        `${expectedBalance} €`
      )
    }
  )

  describe.each`
    description        | traineePrice | session1Price | session2Price
    ${'default price'} | ${10}        | ${undefined}  | ${undefined}
    ${'custom price'}  | ${5}         | ${5}          | ${15}
    ${'mixed price'}   | ${10}        | ${undefined}  | ${20}
  `('with $description', ({traineePrice, session1Price, session2Price}) => {
    test.each`
      description     | session1PayedAmount | session2PayedAmount
      ${'balanced'}   | ${10}               | ${10}
      ${'underpayed'} | ${10}               | ${5}
      ${'overpayed'}  | ${15}               | ${15}
    `('$description', ({session1PayedAmount, session2PayedAmount}) => {
      const price1 = session1Price !== undefined ? session1Price : traineePrice
      const price2 = session2Price !== undefined ? session2Price : traineePrice
      const price = price1 + price2
      const payed = session1PayedAmount + session2PayedAmount
      const expectedBalance = payed - price

      const trainee = new TraineeBuilder().withPrice(traineePrice).build()
      const session1 = new SessionBuilder(trainee.id)
        .withId('Session1')
        .withPrice(session1Price)
        .withPayedAmount(session1PayedAmount)
        .build()
      const session2 = new SessionBuilder(trainee.id)
        .withId('Session2')
        .withPrice(session2Price)
        .withPayedAmount(session2PayedAmount)
        .build()
      trainee.sessionsRef = [session1.id, session2.id]
      // make sure the sessions have different ids
      expect(session1.id).not.toEqual(session2.id)

      const store = createMockStore({
        ...initialApplicationState,
        trainees: {
          [trainee.id]: trainee,
        },
        sessions: {[session1.id]: session1, [session2.id]: session2},
      })

      const {queryByText} = render(
        <Provider store={store}>
          <TraineeBalanceConnected traineeId={trainee.id} />
        </Provider>
      )

      expect(queryByText(/Kontostand/)!.firstElementChild!.textContent).toEqual(
        `${expectedBalance} €`
      )
    })
  })
})
