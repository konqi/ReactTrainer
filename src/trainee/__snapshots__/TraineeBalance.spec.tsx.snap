// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot tests renders properly with negative balance 1`] = `
<body>
  <div>
    <div
      class="MuiPaper-root MuiPaper-elevation1 makeStyles-root-68 MuiPaper-rounded"
    >
      <div
        class="MuiContainer-root MuiContainer-maxWidthLg"
      >
        <h5
          class="MuiTypography-root MuiTypography-h5"
        >
          Kontostand: 
          <span
            class="makeStyles-negativeBalance-69"
          >
            -1
             €
          </span>
        </h5>
      </div>
    </div>
  </div>
</body>
`;

exports[`snapshot tests renders properly with positive balance 1`] = `
<body>
  <div>
    <div
      class="MuiPaper-root MuiPaper-elevation1 makeStyles-root-1 MuiPaper-rounded"
    >
      <div
        class="MuiContainer-root MuiContainer-maxWidthLg"
      >
        <h5
          class="MuiTypography-root MuiTypography-h5"
        >
          Kontostand: 
          <span
            class="makeStyles-positiveBalance-3"
          >
            1
             €
          </span>
        </h5>
      </div>
    </div>
  </div>
</body>
`;

exports[`snapshot tests renders properly with zero balance 1`] = `
<body>
  <div>
    <div
      class="MuiPaper-root MuiPaper-elevation1 makeStyles-root-135 MuiPaper-rounded"
    >
      <div
        class="MuiContainer-root MuiContainer-maxWidthLg"
      >
        <h5
          class="MuiTypography-root MuiTypography-h5"
        >
          Kontostand: 
          <span>
            0
             €
          </span>
        </h5>
      </div>
    </div>
  </div>
</body>
`;
